### PR TITLE
Tests for AddLayerToMesh

### PR DESCRIPTION
**Pull request by rinkk**

### CODE REVIEW

**Comment by rinkk** [comment]
Tests for triangles and quads fail on Jenkins because element flipping is not merged yet.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+		ElementErrorFlag::NonCoplanar, ElementErrorFlag::NonConvex,  ElementErrorFlag::NodeOrder};
+		std::vector<ElementErrorCode> codes (MeshLib::MeshValidation::testElementGeometry(mesh));
+		for (std::size_t i=0; i<codes.size(); ++i)
+			for (std::size_t j=0; j<nErrorFlags-reduce_tests; ++j)
+				ASSERT_EQ(false, codes[i][flags[j]]);
+	}
+
+	void testZCoords2D(MeshLib::Mesh const& input, MeshLib::Mesh const& output, int height)
+	{
+		std::size_t nNodes (input.getNNodes());
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Variable `nNodes` can be `const`.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+#include "MeshLib/MeshEditing/AddLayerToMesh.h"
+#include "MeshLib/MeshQuality/MeshValidation.h"
+
+namespace AddLayerValidation
+{
+	// validates mesh. for line meshes, node order tests will fail because vertical elements are
+	// created during the extrusion. Therefore node order tests are switched off for line meshes.
+	void validate(MeshLib::Mesh const& mesh, bool testNodeOrder)
+	{
+		int reduce_tests = (testNodeOrder) ? 0 : 1;
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Variable `reduce_tests` can be `const`.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+		const std::size_t nErrorFlags (static_cast<std::size_t>(ElementErrorFlag::MaxValue));
+		ElementErrorFlag flags[nErrorFlags] = {ElementErrorFlag::ZeroVolume,
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Can `flags` be `const`?

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			ASSERT_EQ((*input.getNode(i))[2] + height, (*output.getNode(i))[2]);
+	}
+};
+
+TEST(MeshLib, AddLayerToLineMesh)
+{
+	std::unique_ptr<MeshLib::Mesh> mesh (MeshLib::MeshGenerator::generateLineMesh(1.0, 5));
+
+	// add top layer
+	int height (1);
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: `height` can/should be a double value.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+		for (std::size_t i=0; i<nNodes; ++i)
+		{
+			ASSERT_EQ((*input.getNode(i))[2], (*output.getNode(i))[2]);
+			ASSERT_EQ((*input.getNode(i))[2] + height, (*output.getNode(nNodes+i))[2]);
+		}
+	}
+
+	void testZCoords3D(MeshLib::Mesh const& input, MeshLib::Mesh const& output, int height)
+	{
+		std::size_t nNodes (input.getNNodes());
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Variable `nNodes` can be `const`.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	std::array<unsigned, 7> n_elems (MeshLib::MeshInformation::getNumberOfElementTypes(*result));
+	ASSERT_EQ(5, n_elems[0]);
+	ASSERT_EQ(5, n_elems[2]);
+
+	AddLayerValidation::testZCoords2D(*mesh, *result, height);
+	AddLayerValidation::validate(*result, false);
+
+	// add bottom layer
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Only a remark: I would split the test into two tests, one for top layer and one for bottom layer.

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+	// add top layer
+	int height (1);
+	std::unique_ptr<MeshLib::Mesh> result (MeshLib::addLayerToMesh(*mesh, height, "mesh", true));
+
+	ASSERT_EQ(2*mesh->getNNodes(), result->getNNodes());
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	std::array<unsigned, 7> n_elems (MeshLib::MeshInformation::getNumberOfElementTypes(*result));
+	ASSERT_EQ(50, n_elems[1]);
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Can you add a comment what is tested here, i.e., what is `n_elems[1]` and `n_elems[6]` stands for?

File: Tests/MeshLib/TestAddLayerToMesh.cpp [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+	// add top layer
+	int height (1);
+	std::unique_ptr<MeshLib::Mesh> result (MeshLib::addLayerToMesh(*mesh, height, "mesh", true));
+
+	ASSERT_EQ(2*mesh->getNNodes(), result->getNNodes());
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	std::array<unsigned, 7> n_elems (MeshLib::MeshInformation::getNumberOfElementTypes(*result));
+	ASSERT_EQ(25, n_elems[2]);
```
Comment (TomFischer) [inline Tests/MeshLib/TestAddLayerToMesh.cpp]: Please comment what is tested here.

**Comment by TomFischer** [comment]
In general I would split each test into two tests, one for top layer and one for bottom layer.

**Comment by TomFischer** [comment]
Waht do you think: Is the `validate` function useful in other mesh tests, too?

**Comment by rinkk** [comment]
validate tests a number of things directly. it touches even more aspects of what may be wrong with a mesh indirectly. so I guess it's a good way to check if a mesh is valid. there are probably still things that might be wrong. also keep in mind that we use a number of conventions that might result in errors although the mesh is perfectly alright. this includes that normals of 2d mesh elements point roughly upwards or that detecting holes currently works via tracking surfaces.

**Comment by endJunction** [comment]
Just a note; those things which are equal in all tests could be moved up into the test fixture reducing the boilerplate code size...

**Comment by TomFischer** [comment]
:+1:

**Pull request merged**